File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/C/d588eeec57d09bf168bf7d63bc1e6cde.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>
#include <stdint.h>
#include <assert.h>
#include <string.h>

#define MAXN 100005
#define MOD 1000000007

typedef struct {
    int first;
    int second;
} Pair;

int n;
int pts[MAXN][2];
bool visit[MAXN * 2];
bool ve[MAXN * 2];
Pair g[MAXN * 2][MAXN];
int g_size[MAXN * 2];
int x_start, y_start, now = 0;

int gcd(int a, int b) {
    while (a != 0) {
        b %= a;
        if (b == 0)
            return a;
        a %= b;
    }
    return b;
}

int inv_mod(int a, int m) {
    int x, y;
    if (m == 0) {
        x = 1;
        y = 0;
        return a;
    }
    int g = inv_mod(m, a % m);
    y -= (a / m) * x;
    return g;
}

int mul_mod(int a, int b, int m) {
    return (int)(((int64_t)a * b) % m);
}

int pow_mod(int a, int p, int m) {
    int r = 1;
    a %= m;
    while (p) {
        if (p & 1) r = mul_mod(r, a, m);
        a = mul_mod(a, a, m);
        p >>= 1;
    }
    return r;
}

Pair dfs(int u) {
    Pair ret = {1, 0};
    visit[u] = true;
    for (int i = 0; i < g_size[u]; i++) {
        Pair vv = g[u][i];
        if (ve[vv.second]) continue;
        ++ret.second;
        ve[vv.second] = true;
        if (visit[vv.first]) continue;
        Pair p = dfs(vv.first);
        ret.first += p.first;
        ret.second += p.second;
    }
    return ret;
}

int main() {
    scanf("%d", &n);
    
    int trans[2][MAXN] = {0};
    for (int i = 0; i < n; i++) {
        scanf("%d %d", &pts[i][0], &pts[i][1]);
        for (int j = 0; j < 2; j++) {
            if (trans[j][pts[i][j]] == 0) {
                trans[j][pts[i][j]] = now + 1;
                now++;
            }
            pts[i][j] = trans[j][pts[i][j]] - 1;
        }
    }

    x_start = now; now += trans[0][0];
    y_start = now; now += trans[1][0];
    assert(now < MAXN * 2);

    for (int i = 0; i < n; i++) {
        int x = x_start + pts[i][0];
        int y = y_start + pts[i][1];
        g[x][g_size[x]].first = y;
        g[x][g_size[x]].second = i;
        g_size[x]++;
        g[y][g_size[y]].first = x;
        g[y][g_size[y]].second = i;
        g_size[y]++;
    }

    int ans = 1;
    for (int i = 0; i < now; i++) {
        if (!visit[i]) {
            Pair p = dfs(i);
            int k = p.first;
            if (p.second < p.first) {
                ans = mul_mod(ans, (pow_mod(2, k, MOD) - 1 + MOD) % MOD);
            } else {
                ans = mul_mod(ans, pow_mod(2, k, MOD));
            }
        }
    }
    printf("%d\n", ans);
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/d588eeec57d09bf168bf7d63bc1e6cde.java
import java.util.*;

public class d588eeec57d09bf168bf7d63bc1e6cdexcodeeval_processed_4000.json {
    static final int MOD = 1000000007;
    static final int maxn = 100005;
    static int n;
    static int[][] pts = new int[maxn][2];
    static boolean[] visit = new boolean[maxn * 2];
    static boolean[] ve = new boolean[maxn * 2];
    static List<Pair<Integer, Integer>>[] g = new ArrayList[maxn * 2];
    static int x_start, y_start, now = 0;

    static {
        for (int i = 0; i < maxn * 2; i++) {
            g[i] = new ArrayList<>();
        }
    }

    static class Pair<A, B> {
        A first;
        B second;

        Pair(A first, B second) {
            this.first = first;
            this.second = second;
        }
    }

    static Pair<Integer, Integer> dfs(int u) {
        Pair<Integer, Integer> ret = new Pair<>(1, 0);
        visit[u] = true;
        for (Pair<Integer, Integer> vv : g[u]) {
            if (ve[vv.second]) continue;
            ret.second++;
            ve[vv.second] = true;
            if (visit[vv.first]) continue;
            Pair<Integer, Integer> p = dfs(vv.first);
            ret.first += p.first;
            ret.second += p.second;
        }
        return ret;
    }

    public static void main(String[] args) {
        Scanner fin = new Scanner(System.in);
        n = fin.nextInt();

        Map<Integer, Integer>[] trans = new HashMap[2];
        trans[0] = new HashMap<>();
        trans[1] = new HashMap<>();

        for (int i = 0; i < n; i++) {
            pts[i][0] = fin.nextInt();
            pts[i][1] = fin.nextInt();
            for (int j = 0; j < 2; j++) {
                if (!trans[j].containsKey(pts[i][j])) {
                    int s = trans[j].size();
                    trans[j].put(pts[i][j], s);
                }
                pts[i][j] = trans[j].get(pts[i][j]);
            }
        }

        x_start = now; now += trans[0].size();
        y_start = now; now += trans[1].size();
        assert now < maxn * 2;

        for (int i = 0; i < n; i++) {
            int x = x_start + pts[i][0];
            int y = y_start + pts[i][1];
            g[x].add(new Pair<>(y, i));
            g[y].add(new Pair<>(x, i));
        }

        int ans = 1;
        for (int i = 0; i < now; i++) {
            if (!visit[i]) {
                Pair<Integer, Integer> p = dfs(i);
                int k = p.first;
                if (p.second < p.first) {
                    ans = (int) ((ans * (modPow(2, k) - 1 + MOD) % MOD) % MOD);
                } else {
                    ans = (int) ((ans * modPow(2, k)) % MOD);
                }
            }
        }
        System.out.println(ans);
    }

    static long modPow(int a, int b) {
        long res = 1;
        long base = a % MOD;
        while (b > 0) {
            if ((b & 1) == 1) {
                res = (res * base) % MOD;
            }
            base = (base * base) % MOD;
            b >>= 1;
        }
        return res;
    }
}
